src/bussinessLogic/ManageArray.java

```diff
@@ -79,14 +94,92 @@ public void setNewCharMap() {
         }
     }
 
-    public LinkedHashMap<String, Integer> getCharMap() {
-        return charMap;
+    @Override
+    // calculates the average time for each char
+    public void calculateAverage() {
+        currentChar = 'a';
+        int average;
+
+        for (int i = 0; i < numberOfLetters; i++) {
+            charString = Character.toString(currentChar);
+            average = charMap.get(charString) / dividerMap.get(charString);
+            charMap.put(charString, average);
+            currentChar++;
+        }
     }
 
-    public LinkedHashMap<String, Integer> getDividerMap() {
-        return dividerMap;
+    @Override
+    // sort the charMap in ascending order
+    public void sortCharMap() {
+        // generate a new, empty sortedMap
+        sortedMap = new LinkedHashMap<>();
+        // arrayList for storing and sorting the chars
+        ArrayList<Integer> charList = new ArrayList<>();
+        // transfer the chars from charMap to charList
+        for (Map.Entry<String, Integer> entry : charMap.entrySet()) {
+            charList.add(entry.getValue());
+        }
+        // rank the chars by typing speed (fastest to slowest)
+        Collections.sort(charList);
+        // transfer the ranked list to sortedMap
+        for (int num : charList) {
+            for (Map.Entry<String, Integer> entry : charMap.entrySet()) {
+                if (entry.getValue().equals(num)) {
+                    sortedMap.put(entry.getKey(), num);
+                }
+            }
+        }
     }
 
+    @Override
+
+    // transfer the sorted chars to a char array
+    public void setCharArrayFromSortedMap() {
+        charArrayFromSortedMap = sortedMap.keySet().stream().map(String::valueOf).collect(Collectors.joining())
+                .toCharArray();
+
+    }
+
+    @Override
+    /*
+     * generate an array with characters repeating proportionally to the reaction
+     * time for each char
+     */
+    public void setPracticeArray() {
+
+        // calculate the size of the practiceArray
+        int sizeOfPracticeArray = numberOfLetters * (numberOfLetters + 1) / 2;
+        // create the array
+        practiceArray = new char[sizeOfPracticeArray];
+        // set index
+        int index = 0;
+
+        for (int i = 0; i < charArrayFromSortedMap.length; i++) {
+            for (int j = 0; j <= i; j++) {
+                practiceArray[index] = charArrayFromSortedMap[i];
+                index++;
+            }
+        }
+    }
+
+    @Override
+    // calculate how many times a char was deployed and store the result in a hashmap
+    public void addDividerToMap(char typedChar) {
```

Comment:
`typedChar` is passed in both `addDividerToMap `and `addTimeToMap `methods. In the next pull request I shall address this problem and pass the parameter to the constructor